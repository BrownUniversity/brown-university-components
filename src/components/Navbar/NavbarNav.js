import PropTypes from 'prop-types';
import React, { Component } from 'react';
import styled from 'styled-components';
import { WindowSize } from 'react-fns';
import { Collapse } from 'react-collapse';

import NavbarContext from './NavbarContext';
import Hamburger from '../Hamburger';
import Nav from '../Nav';
import breakpoints from '../../constants/breakpoints';
import colors from '../../constants/colors';
import media from '../../constants/media';

/*
  inner components
*/
// filter props so they don't become dom attributes (see `styled-components` issue 439)
const MobileCollapseWrapper = styled(({ color, ...restProps }) => (
  <div {...restProps} />
))`
  box-shadow: 0 5px 10px 0 #00000026;
  left: 0;
  position: absolute;
  top: 75px;
  width: 100%;
  background-color: ${({ color }) => colors[color]};
`;

const MobileNavWrapper = styled.div`
  padding: 0 7vw 1rem;

  ${media.md`
    padding: 0 33px;
  `};
`;

/*
  outer NavbarNav component
*/
const getChildColor = color => {
  switch (color) {
    case 'white':
      return 'red';

    // brown
    default:
      return 'white';
  }
};

class NavbarNav extends Component {
  state = {
    mobileNavIsOpen: false
  };

  handleMobileNavToggle = () =>
    this.setState(({ mobileNavIsOpen }) => ({
      mobileNavIsOpen: !mobileNavIsOpen
    }));

  render() {
    const { mobileBreakpoint, ...restProps } = this.props;
    const { mobileNavIsOpen } = this.state;

    return (
      <WindowSize
        render={({ width }) => {
          // TODO: update when `width` doesn't return 0 on initial render (see `react-fns` issue 84)
          const currentWidth = width === 0 ? window.innerWidth : width;
          const renderMobile = currentWidth < mobileBreakpoint;

          return (
            <NavbarContext.Consumer>
              {({ color }) => {
                const childColor = getChildColor(color);

                if (renderMobile) {
                  return (
                    <div {...restProps}>
                      <Hamburger
                        aria-controls="navbar-nav-mobile-collapse"
                        color={childColor}
                        isOpen={mobileNavIsOpen}
                        onOpen={this.handleMobileNavToggle}
                        onClose={this.handleMobileNavToggle}
                      />
                      <MobileCollapseWrapper color={color}>
                        <Collapse
                          id="navbar-nav-mobile-collapse"
                          isOpened={mobileNavIsOpen}
                        >
                          <MobileNavWrapper>
                            <Nav mobile color={childColor}>
                              {this.props.children}
                            </Nav>
                          </MobileNavWrapper>
                        </Collapse>
                      </MobileCollapseWrapper>
                    </div>
                  );
                }

                return (
                  <Nav {...restProps} navbar color={childColor}>
                    {this.props.children}
                  </Nav>
                );
              }}
            </NavbarContext.Consumer>
          );
        }}
      />
    );
  }
}

NavbarNav.propTypes = {
  mobileBreakpoint: PropTypes.number,
  children: PropTypes.node.isRequired
};

NavbarNav.defaultProps = {
  mobileBreakpoint: breakpoints.md
};

NavbarNav.Item = Nav.Item;
NavbarNav.Link = Nav.Link;

export default NavbarNav;
