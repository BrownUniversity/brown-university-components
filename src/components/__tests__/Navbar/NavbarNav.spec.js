import React from "react";
import { breakpoints } from "brown-university-styles";
import { render, fireEvent } from "@testing-library/react";
import { triggerWindowResize, resetWindowSize } from "window-test-utils";
import NavbarContext from "../../Navbar/NavbarContext";
import NavbarNav from "../../Navbar/NavbarNav";

const renderNavbarNav = ({
  context = {},
  props = {},
  children = (
    <React.Fragment>
      <NavbarNav.Item>
        <NavbarNav.Link>Link</NavbarNav.Link>
      </NavbarNav.Item>
      <NavbarNav.Item>
        <NavbarNav.Link active>Active Link</NavbarNav.Link>
      </NavbarNav.Item>
      <NavbarNav.Item>
        <NavbarNav.Link disabled>Link</NavbarNav.Link>
      </NavbarNav.Item>
    </React.Fragment>
  ),
} = {}) => {
  const { color = "brown", mobileBreakpoint = breakpoints.md } = context;
  const rtlUtils = render(
    <NavbarContext.Provider value={{ color, mobileBreakpoint }}>
      <NavbarNav {...props}>{children}</NavbarNav>
    </NavbarContext.Provider>,
  );
  const navbarNav = rtlUtils.container.firstChild;
  return {
    navbarNav,
    navbarNavMobileNavWrapper: navbarNav.children[1],
    ...rtlUtils,
  };
};

describe("NavbarNav", () => {
  describe("when window width is above mobileBreakpoint", () => {
    it("should render default", () => {
      const { navbarNav } = renderNavbarNav();

      expect(navbarNav).toMatchSnapshot();
    });
  });

  describe("when window width is below mobileBreakpoint", () => {
    beforeEach(() => {
      triggerWindowResize({ width: breakpoints.md - 1 });
    });

    afterAll(() => {
      resetWindowSize();
    });

    describe("styles", () => {
      it("should render brown mobile navbar nav by default", () => {
        const { navbarNavMobileNavWrapper } = renderNavbarNav();

        expect(navbarNavMobileNavWrapper).toMatchSnapshot();
      });

      it("should render white mobile navbar nav when color variant is provided", () => {
        const { navbarNavMobileNavWrapper } = renderNavbarNav({
          context: { color: "white" },
        });

        expect(navbarNavMobileNavWrapper).toMatchSnapshot();
      });
    });

    it("should toggle navbar nav collapse", () => {
      const { getByLabelText } = renderNavbarNav();
      const toggleButtonByLabelText = getByLabelText("Toggle navigation");

      expect(toggleButtonByLabelText).toHaveAttribute("aria-expanded", "false");
      fireEvent.click(toggleButtonByLabelText);
      expect(toggleButtonByLabelText).toHaveAttribute("aria-expanded", "true");
      fireEvent.click(toggleButtonByLabelText);
      expect(toggleButtonByLabelText).toHaveAttribute("aria-expanded", "false");
    });
  });
});
