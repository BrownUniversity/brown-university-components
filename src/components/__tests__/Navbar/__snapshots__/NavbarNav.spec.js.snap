// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavbarNav when window width is above mobileBreakpoint should render default 1`] = `
.c1 {
  display: list-item;
  padding: 9px 15px;
}

.c2 {
  background: transparent;
  border: none;
  display: block;
  font-family: CircularStd,Arial,Helvetica,sans-serif;
  padding: 0;
  -webkit-text-decoration: none !important;
  text-decoration: none !important;
  color: #FFFFFF;
  cursor: pointer;
  font-size: 1em;
  font-weight: 400;
  line-height: 1.5;
  opacity: 1;
  pointer-events: auto;
}

.c2::after {
  content: "";
  display: block;
  height: 3px;
  background: #CCC6B8;
  margin-top: 0.25em;
  width: 0;
}

.c2:hover {
  color: #CCC6B8;
}

.c2:hover::after {
  width: 100%;
}

.c3 {
  background: transparent;
  border: none;
  display: block;
  font-family: CircularStd,Arial,Helvetica,sans-serif;
  padding: 0;
  -webkit-text-decoration: none !important;
  text-decoration: none !important;
  color: #CCC6B8;
  cursor: pointer;
  font-size: 1em;
  font-weight: 400;
  line-height: 1.5;
  opacity: 1;
  pointer-events: auto;
}

.c3::after {
  content: "";
  display: block;
  height: 3px;
  background: #CCC6B8;
  margin-top: 0.25em;
  width: 100%;
}

.c3:hover {
  color: #CCC6B8;
}

.c3:hover::after {
  width: 100%;
}

.c4 {
  background: transparent;
  border: none;
  display: block;
  font-family: CircularStd,Arial,Helvetica,sans-serif;
  padding: 0;
  -webkit-text-decoration: none !important;
  text-decoration: none !important;
  color: #FFFFFF;
  cursor: not-allowed;
  font-size: 1em;
  font-weight: 400;
  line-height: 1.5;
  opacity: 1;
  pointer-events: auto;
}

.c4::after {
  content: "";
  display: block;
  height: 3px;
  background: #CCC6B8;
  margin-top: 0.25em;
  width: 0;
}

.c4:hover {
  color: #FFFFFF;
}

.c4:hover::after {
  width: 0;
}

.c0 {
  list-style: none;
  margin: 0;
  background-color: transparent;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0;
  width: auto;
}

<ul
  class="c0"
>
  <li
    class="c1"
  >
    <button
      class="c2"
      color="white"
    >
      Link
    </button>
  </li>
  <li
    class="c1"
  >
    <button
      class="c3"
      color="white"
    >
      Active Link
    </button>
  </li>
  <li
    class="c1"
  >
    <button
      class="c4"
      color="white"
      disabled=""
    >
      Link
    </button>
  </li>
</ul>
`;

exports[`NavbarNav when window width is below mobileBreakpoint styles should render brown mobile navbar nav by default 1`] = `
.c0 {
  background: transparent;
  border: none;
  cursor: pointer;
  height: 24px;
  padding: 0 25px 3px 0;
}

.c1 {
  height: 3px;
  position: absolute;
  width: 25px;
  background-color: #FFFFFF;
  -webkit-transition-property: -webkit-transform;
  -webkit-transition-property: transform;
  transition-property: transform;
  -webkit-transition-timing-function: ease;
  transition-timing-function: ease;
  -webkit-transition-duration: .22s;
  transition-duration: .22s;
  -webkit-transition-timing-function: cubic-bezier(0.55,0.055,0.675,0.19);
  transition-timing-function: cubic-bezier(0.55,0.055,0.675,0.19);
}

.c1.c1:before {
  height: 3px;
  position: absolute;
  width: 25px;
  background-color: #FFFFFF;
  content: "";
  display: block;
  -webkit-transition-property: -webkit-transform;
  -webkit-transition-property: transform;
  transition-property: transform;
  -webkit-transition-timing-function: ease;
  transition-timing-function: ease;
  -webkit-transition-duration: .15s;
  transition-duration: .15s;
  -webkit-transition: top .1s .25s ease-in,opacity .1s ease-in;
  transition: top .1s .25s ease-in,opacity .1s ease-in;
  top: -7px;
}

.c1.c1:after {
  height: 3px;
  position: absolute;
  width: 25px;
  background-color: #FFFFFF;
  content: "";
  display: block;
  -webkit-transition-property: -webkit-transform;
  -webkit-transition-property: transform;
  transition-property: transform;
  -webkit-transition-timing-function: ease;
  transition-timing-function: ease;
  -webkit-transition-duration: .15s;
  transition-duration: .15s;
  -webkit-transition: bottom .1s .25s ease-in,-webkit-transform .22s cubic-bezier(0.55,0.055,0.675,0.19);
  -webkit-transition: bottom .1s .25s ease-in,transform .22s cubic-bezier(0.55,0.055,0.675,0.19);
  transition: bottom .1s .25s ease-in,transform .22s cubic-bezier(0.55,0.055,0.675,0.19);
  bottom: -7px;
}

.c6 {
  display: list-item;
  border-top: 0.5px solid #C8C8C880;
  padding: 1rem 0;
}

.c7 {
  background: transparent;
  border: none;
  display: block;
  font-family: CircularStd,Arial,Helvetica,sans-serif;
  padding: 0;
  -webkit-text-decoration: none !important;
  text-decoration: none !important;
  color: #FFFFFF;
  cursor: pointer;
  font-size: 1em;
  font-weight: 700;
  line-height: 1.5;
  opacity: 1;
  pointer-events: auto;
  -webkit-transition: all .2s ease-in-out;
  transition: all .2s ease-in-out;
}

.c7:hover {
  color: #CCC6B8;
}

.c8 {
  background: transparent;
  border: none;
  display: block;
  font-family: CircularStd,Arial,Helvetica,sans-serif;
  padding: 0;
  -webkit-text-decoration: none !important;
  text-decoration: none !important;
  color: #CCC6B8;
  cursor: pointer;
  font-size: 1em;
  font-weight: 700;
  line-height: 1.5;
  opacity: 1;
  pointer-events: auto;
  -webkit-transition: all .2s ease-in-out;
  transition: all .2s ease-in-out;
}

.c8:hover {
  color: #CCC6B8;
}

.c9 {
  background: transparent;
  border: none;
  display: block;
  font-family: CircularStd,Arial,Helvetica,sans-serif;
  padding: 0;
  -webkit-text-decoration: none !important;
  text-decoration: none !important;
  color: #FFFFFF;
  cursor: not-allowed;
  font-size: 1em;
  font-weight: 700;
  line-height: 1.5;
  opacity: 1;
  pointer-events: auto;
  -webkit-transition: all .2s ease-in-out;
  transition: all .2s ease-in-out;
}

.c9:hover {
  color: #FFFFFF;
}

.c5 {
  list-style: none;
  margin: 0;
  background-color: transparent;
  display: block;
  padding: 0;
  width: 100%;
}

.c3 {
  will-change: height;
  display: none;
  height: 0;
  overflow: hidden;
  -webkit-transition: height 250ms cubic-bezier(0.4,0,0.2,1);
  transition: height 250ms cubic-bezier(0.4,0,0.2,1);
}

.c2 {
  box-shadow: 0 5px 10px 0 #00000026;
  left: 0;
  position: absolute;
  top: 75px;
  width: 100%;
  background-color: #4E3629;
}

.c4 {
  padding: 0 7vw 1rem;
}

@media (min-width:740px) {
  .c4 {
    padding: 0 33px;
  }
}

<div>
  <button
    aria-controls="navbar-nav-mobile-collapse"
    aria-expanded="false"
    aria-label="Toggle navigation"
    class="c0"
    type="button"
  >
    <span
      class="c1"
    />
  </button>
  <div
    class="c2"
  >
    <div
      class="c3"
      display="none"
      height="0"
      id="navbar-nav-mobile-collapse"
      overflow="hidden"
    >
      <div
        class="c4"
      >
        <ul
          class="c5"
        >
          <li
            class="c6"
          >
            <button
              class="c7"
              color="white"
            >
              Link
            </button>
          </li>
          <li
            class="c6"
          >
            <button
              class="c8"
              color="white"
            >
              Active Link
            </button>
          </li>
          <li
            class="c6"
          >
            <button
              class="c9"
              color="white"
              disabled=""
            >
              Link
            </button>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
